airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/ConfigPersistenceFactory.java

```diff
@@ -0,0 +1,151 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * By default, this factory returns a database config persistence. it can still return a file system
+ * config persistence for testing purpose. This legacy feature should be removed after the file to
+ * database migration is completely done.
+ */
+public class ConfigPersistenceFactory {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ConfigPersistenceFactory.class);
+
+  private final Configs configs;
+  private final boolean setupDatabase;
+  private final boolean useConfigDatabase;
+
+  /**
+   * @param setupDatabase initialize the database and load data; this is necessary because this method
+   *        has multiple callers, and we want to setup the database only once to prevent race
+   *        conditions.
+   */
+  private ConfigPersistenceFactory(Configs configs, boolean setupDatabase, boolean useConfigDatabase) {
+    this.configs = configs;
+    this.setupDatabase = setupDatabase;
+    this.useConfigDatabase = useConfigDatabase;
+  }
+
+  public static Builder build(Configs configs) {
+    return new Builder(configs);
+  }
+
+  public static class Builder {
+
+    private final Configs configs;
+    private boolean setupDatabase = false;
+    private boolean useConfigDatabase = true;
+
+    private Builder(Configs configs) {
+      this.configs = configs;
+    }
+
+    public Builder setupDatabase(boolean setupDatabase) {
+      this.setupDatabase = setupDatabase;
+      return this;
+    }
+
+    public Builder useConfigDatabase(boolean useConfigDatabase) {
+      this.useConfigDatabase = useConfigDatabase;
+      return this;
+    }
+
+    public ConfigPersistenceFactory get() {
+      return new ConfigPersistenceFactory(configs, setupDatabase, useConfigDatabase);
+    }
+
+  }
+
+  /**
+   * Create a config persistence based on the configs.
+   * <p/>
+   * If config root is defined, create a database config persistence and copy the configs from the
+   * file-based config persistence. Otherwise, seed the database from the yaml files.
+   */
+  public ConfigPersistence create() throws IOException {
+    if (!useConfigDatabase) {
+      Path configRoot = configs.getConfigRoot();
+      LOGGER.info("Use file system config persistence (root: {})", configRoot);
+      return FileSystemConfigPersistence.createWithValidation(configRoot);
+    }
+
+    if (configs.getConfigRoot() == null) {
+      // This branch will only be true in a future Airbyte version, in which
+      // the config root is no longer required and everything lives in the database.
+      return createDbPersistenceWithYamlSeed();
+    }
+
+    return createDbPersistenceWithFileSeed();
+  }
+
+  ConfigPersistence createDbPersistenceWithYamlSeed() throws IOException {
+    ConfigPersistence seedConfigPersistence = new YamlSeedConfigPersistence();
+    return createDbPersistence(seedConfigPersistence);
+  }
+
+  ConfigPersistence createDbPersistenceWithFileSeed() throws IOException {
```

Comment:
this method is only for when we create a new PG database seeding it with an existing File based config database, right? maybe a comment to clarify?

Reply:
Added comments.